pkg/relayer/proxy/relay_meter.go

```diff
@@ -0,0 +1,384 @@
+	if block.Height()%numBlocksPerSession == 0 {
+		rmtr.supplierToAppMetricsMap = make(map[string]map[string]*appRelayMeter)
+	}
+}
+
+// forEachEventApplicationStakedFn is a callback function that is called every time
+// an application staked event is observed. It updates the relay meter known applications.
+func (rmtr *ProxyRelayMeter) forEachEventApplicationStakedFn(ctx context.Context, event *apptypes.EventApplicationStaked) {
+	rmtr.relayMeterMu.Lock()
+	defer rmtr.relayMeterMu.Unlock()
+
+	app := event.GetApplication()
+
+	// Since lean clients are supported, multiple suppliers might share the same RelayMiner.
+	// Loop over all the suppliers that have metered the application and update their
+	// max amount of stake they can consume.
+	for supplierAddress := range rmtr.supplierToAppMetricsMap {
+		appMetrics, ok := rmtr.supplierToAppMetricsMap[supplierAddress][app.GetAddress()]
+		if !ok {
+			continue
+		}
+		appMetrics.app.Stake = app.GetStake()
+		appStakeShare := getSupplierAppStake(app.GetStake(), appMetrics.sharedParams)
+		appMetrics.maxCoin = appStakeShare.Add(rmtr.overServicingAllowanceCoins)
+	}
+}
+
+func (rmtr *ProxyRelayMeter) ensureRequestAppMetrics(ctx context.Context, reqMeta servicetypes.RelayRequestMetadata, createMetric bool) (*appRelayMeter, error) {
+	appAddress := reqMeta.GetSessionHeader().GetApplicationAddress()
+	supplierAddress := reqMeta.GetSupplierOperatorAddress()
+
+	supplierApps, ok := rmtr.supplierToAppMetricsMap[supplierAddress]
+	if !ok {
+		rmtr.supplierToAppMetricsMap[supplierAddress] = make(map[string]*appRelayMeter)
+		supplierApps = rmtr.supplierToAppMetricsMap[supplierAddress]
+	}
+
+	// Do not consider applications that have not been seen in this session.
+	appMetrics, ok := supplierApps[appAddress]
+
+	// If the application is seen for the first time in this session, calculate the
+	// max amount of stake the application can consume.
+	if !ok {
+		var app apptypes.Application
+		app, err := rmtr.applicationQuerier.GetApplication(ctx, appAddress)
+		if err != nil {
+			return nil, err
+		}
+
+		sharedParams, err := rmtr.sharedQuerier.GetParams(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		service, err := rmtr.serviceQuerier.GetService(ctx, reqMeta.SessionHeader.ServiceId)
+		if err != nil {
+			return nil, err
+		}
+
+		serviceRelayDifficulty, err := rmtr.serviceQuerier.GetServiceRelayDifficulty(ctx, service.Id)
+		if err != nil {
+			return nil, err
+		}
+
+		if !createMetric {
+			return nil, ErrRelayerProxyUnknownApplication.Wrap("required metric not found")
+		}
+
+		// calculate the max amount of stake the application can consume in the current session.
+		supplierAppStake := getSupplierAppStake(app.Stake, sharedParams)
+		maxAmount := supplierAppStake.Add(rmtr.overServicingAllowanceCoins)
+		appMetrics = &appRelayMeter{
+			app:                    app,
+			consumedCoin:           cosmostypes.NewInt64Coin(volatile.DenomuPOKT, 0),
+			maxCoin:                maxAmount,
```

Comment:
Why isn't this `AppStake/NumSuppliersPerSession` w/ the assumption (comment) that the app is staked for 1 service only (can we enforce this?)

Reply:
It was including the over servicing allowance but I've updated it.